src/listpack.c

```diff
@@ -997,6 +1022,11 @@ int lpValidateNext(unsigned char *lp, unsigned char **pp, size_t lpbytes) {
 #undef OUT_OF_RANGE
 }
 
+/* Validate that the entry doesn't reach outside the listpack allocation. */
+static inline void lpAssertValidEntry(unsigned char* lp, size_t lpbytes, unsigned char *p) {
+    assert(lpValidateNext(lp, &p, lpbytes));
+}
```

Comment:
1. why have this and not use the `ASSERT_INTEGRITY` macro? the macro is in theory better since it is sure to get inlined (the `inline` is just a recommendation that the compiler can ignore).
2. why use `lpValidateNext` (also moves to the next element, i.e. returns a new `p` which we don't use)

Reply:
`ASSERT_INTEGRITY` does not apply, because when entry_size exceeds, `ASSERT_INTEGRITY` does not trigger the assertion, see my new `corrupt payload: fuzzer findings - hash listpack too long entry len` test , which passes the verification in the old code.

Reply:
you mean that it's because it doesn't validate the prevlen part?
out of the 3 places that use it, that's probably only relevant for `lpPrev`, and even there, arguably it is done too late.
if that's the case, i'd suggest to revert the other places and improve `lpPrev` to assert on the result of lpDecodeBacklen or the range it leads to.
but maybe the existing call to `ASSERT_INTEGRITY` after adjusting `p` already covers that.

maybe i'm missing something, or maybe i just need to run that new test you added and debug it.. let me know.

Reply:
The stack of `corrupt payload: fuzzer findings - hash listpack too long entry len`.

```
src/redis-server 127.0.0.1:21159(_addReplyProtoToList+0xd3)[0x5663c8a3]
src/redis-server 127.0.0.1:21159(+0x538ff)[0x5663c8ff]
src/redis-server 127.0.0.1:21159(addReplyBulkCBuffer+0x39)[0x5663e709]
src/redis-server 127.0.0.1:21159(+0x8885c)[0x5667185c]
src/redis-server 127.0.0.1:21159(genericHgetallCommand+0x11b)[0x5667335b]
src/redis-server 127.0.0.1:21159(hvalsCommand+0x12)[0x56673412]
```

Because this command copies data while iterating, if the entry_size is wrong, `lpGet` will read an illegal address.

Reply:
I use lpValidateNext simply because it verifies that the entry_size is not out of bounds.
I don't want it to move to the next entry.

Reply:
so the violation in this case was because of lpGet (after lpNext)?
maybe we need to add a call to ASSERT_INTEGRITY_LEN in lpGet?

> I use lpValidateNext simply because it verifies that the entry_size is not out of bounds.

maybe add a comment explaining it.

Reply:
>  so the violation in this case was because of lpGet (after lpNext)?

Yes.

> maybe we need to add a call to ASSERT_INTEGRITY_LEN in lpGet?

I thought about it, but `lpGet` doesn't have `lp` variable, so `ASSERT_INTEGRITY_LEN` can't be used in `lpGet`.

Reply:
actually, i wanna go back on that.. the contract i was trying to make is that when some function returns a `p`, it should validate that `p`, and when a function takes a `p` as input, it can assume it is safe to use.

Reply:
That's it!